demos/signature/EVP_ED_Signature_demo.c

```diff
@@ -0,0 +1,210 @@
+/*-
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * This demonstration will calculate and verify an ED25519 signature of
+ * a message using  EVP_DigestSign() and EVP_DigestVerify().
+ */
+
+#include <string.h>
+#include <stdio.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/core_names.h>
+
+/* A test message to sign */
+static const unsigned char hamlet[] =
+    "To be, or not to be, that is the question,\n"
+    "Whether tis nobler in the minde to suffer\n"
+    "The slings and arrowes of outragious fortune,\n"
+    "Or to take Armes again in a sea of troubles,\n";
+
+static int demo_sign(EVP_PKEY *priv,
+                     const unsigned char *tbs, size_t tbs_len,
+                     OSSL_LIB_CTX *libctx,
+                     unsigned char **sig_out_value,
+                     size_t *sig_out_len)
+{
+    int ret = 0;
+    size_t sig_len;
+    unsigned char *sig_value = NULL;
+    EVP_MD_CTX *sign_context = NULL;
+
+    /* Create a signature context */
+    sign_context = EVP_MD_CTX_new();
+    if (sign_context == NULL) {
+        fprintf(stderr, "EVP_MD_CTX_new failed.\n");
+        goto cleanup;
+    }
+
+    /*
+     * Initialize the sign context using a ED25519 private key
+     * Notice that the digest name must NOT be used.
+     * In this demo we don't specify any additional parameters via
+     * OSSL_PARAM, which means it will use default values.
+     * For more information, refer to
+     * doc/man7/EVP_SIGNATURE-ED25519.pod
+     * "ED25519 and ED448 Signature Parameters"
+     */
+    if (!EVP_DigestSignInit_ex(sign_context, NULL, NULL, libctx, NULL, priv, NULL)) {
+        fprintf(stderr, "EVP_DigestSignInit_ex failed.\n");
+        goto cleanup;
+    }
+
+    /* Calculate the required size for the signature by passing a NULL buffer. */
+    if (!EVP_DigestSign(sign_context, NULL, &sig_len, tbs, tbs_len)) {
+        fprintf(stderr, "EVP_DigestSign using NULL buffer failed.\n");
+        goto cleanup;
+    }
+    sig_value = OPENSSL_zalloc(sig_len);
```

Comment:
Why zalloc here and not malloc?

Reply:
I guess it is a choice between having a zero'd value and something potentially random when it goes wrong.